crates/accelerate/src/circuit_library/iqp.rs

```diff
@@ -0,0 +1,164 @@
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use std::f64::consts::PI;
+
+use ndarray::{Array2, ArrayView2};
+use numpy::PyReadonlyArray2;
+use pyo3::prelude::*;
+use qiskit_circuit::{
+    circuit_data::CircuitData,
+    operations::{Param, StandardGate},
+    Qubit,
+};
+use rand::{Rng, SeedableRng};
+use rand_pcg::Pcg64Mcg;
+use smallvec::{smallvec, SmallVec};
+
+use crate::CircuitError;
+
+const PI2: f64 = PI / 2.0;
+const PI8: f64 = PI / 8.0;
+
+fn iqp(
+    interactions: ArrayView2<i64>,
+) -> impl Iterator<Item = (StandardGate, SmallVec<[Param; 3]>, SmallVec<[Qubit; 2]>)> + '_ {
+    let num_qubits = interactions.ncols();
+
+    // The initial and final Hadamard layer.
+    let h_layer =
+        (0..num_qubits).map(|i| (StandardGate::HGate, smallvec![], smallvec![Qubit(i as u32)]));
+
+    // The circuit interactions are powers of the CSGate, which is implemented by calling
+    // the CPhaseGate with angles of Pi/2 times the power. The gate powers are given by the
+    // upper triangular part of the symmetric ``interactions`` matrix.
+    let connections = (0..num_qubits).flat_map(move |i| {
+        (i + 1..num_qubits)
+            .map(move |j| (j, interactions[(i, j)]))
+            .filter(move |(_, value)| value % 4 != 0)
+            .map(move |(j, value)| {
+                (
+                    StandardGate::CPhaseGate,
+                    smallvec![Param::Float(PI2 * value as f64)],
+                    smallvec![Qubit(i as u32), Qubit(j as u32)],
+                )
+            })
+    });
+
+    // The layer of T gates. Again we use the PhaseGate, now with powers of Pi/8. The powers
+    // are given by the diagonal of the ``interactions`` matrix.
+    let shifts = (0..num_qubits)
+        .map(move |i| interactions[(i, i)])
+        .enumerate()
+        .filter(|(_, value)| value % 8 != 0)
+        .map(|(i, value)| {
+            (
+                StandardGate::PhaseGate,
+                smallvec![Param::Float(PI8 * value as f64)],
+                smallvec![Qubit(i as u32)],
+            )
+        });
+
+    h_layer
+        .clone()
+        .chain(connections)
+        .chain(shifts)
+        .chain(h_layer)
+}
+
+/// This generates a random symmetric integer matrix with values in [0,7].
+fn generate_random_interactions(num_qubits: u32, seed: Option<u64>) -> Array2<i64> {
+    let num_qubits = num_qubits as usize;
+    let mut rng = match seed {
+        Some(seed) => Pcg64Mcg::seed_from_u64(seed),
+        None => Pcg64Mcg::from_entropy(),
+    };
+
+    let mut mat = Array2::zeros((num_qubits, num_qubits));
+    for i in 0..num_qubits {
+        for j in 0..i {
+            mat[[i, j]] = rng.gen_range(0..8) as i64;
+            mat[[j, i]] = mat[[i, j]];
+        }
+    }
```

Comment:
As discussed with @Cryoris, we should also allow producing non-zero elements on the diagonal.